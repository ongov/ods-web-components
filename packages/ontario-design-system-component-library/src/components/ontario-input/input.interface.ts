import { Input } from '../../utils/common.interface';
import { Label } from '../../utils/label/label.interface';

export interface TextInput extends Input, Label {
	/**
	 * The width of the input field. If no value is assigned, it will present as the default input width.
	 */
	inputWidth?: string;

	/**
	 * The input type value.
	 */
	type?: string;
}
