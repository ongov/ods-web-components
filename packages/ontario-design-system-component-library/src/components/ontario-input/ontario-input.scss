@use '../../utils/input-caption/input-caption.scss';
@use '../../global.scss';

$ontario-input-padding: 0.625rem;

.ontario-input {
	border: global.$border-size-standard solid global.$ontario-colour-black;
	border-radius: global.$global-radius;
	box-sizing: border-box;
	color: global.$ontario-colour-black;
	display: block;
	font-size: 1rem;
	font-family: global.$ontario-font-open-sans;
	line-height: global.$line-height-default;
	margin: global.$spacing-0 global.$spacing-0 global.$spacing-7;
	max-width: global.$standard-width;
	padding: $ontario-input-padding global.$spacing-4;
	transition: global.$ontario-focus-transition;
	width: global.$full-width;

	&:focus,
	&:active {
		@extend %ontario-focus !optional;
	}
}

.ontario-input-hint-expander--true {
	margin-bottom: global.$spacing-4;
}

.ontario-input--2-char-width {
	max-width: 8ex;
}

.ontario-input--3-char-width {
	max-width: 10ex;
}

.ontario-input--4-char-width {
	max-width: 12ex;
}

.ontario-input--5-char-width {
	max-width: 14ex;
}

.ontario-input--7-char-width {
	max-width: 17ex;
}

.ontario-input--10-char-width {
	max-width: 23ex;
}

.ontario-input--20-char-width {
	max-width: 41ex;
}
