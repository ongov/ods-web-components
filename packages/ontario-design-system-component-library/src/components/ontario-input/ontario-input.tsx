import { Component, Event, EventEmitter, h, Prop, State } from '@stencil/core';
import { v4 as uuid } from 'uuid';
import { TextInput } from './input.interface';
import { getLabelElement } from '../../utils/label/label';

/**
 * Ontario Input component
 */
@Component({
	tag: 'ontario-input',
	styleUrl: 'ontario-input.scss',
	shadow: true,
})
export class OntarioInput implements TextInput {
	/**
	 * The text to display as label.
	 */
	@Prop() labelCaption: string;

	/**
	 * The form control with which the caption is associated.
	 */
	@Prop({ mutable: true }) labelFor: string;

	/**
	 * The type of label to render.
	 */
	@Prop() labelType: 'default' | 'large' | 'heading' = 'default';

	/**
	 * The aria-describedBy value if the input has hint text associated with it.
	 */
	@Prop() describedBy?: string;

	/**
	 * The unique identifier of the input. If no ID is passed, one will be autogenerated.
	 */
	@Prop({ mutable: true }) elementId?: string;

	/**
	 * The width of the input field. If no value is assigned, it will present as the default input width.
	 */
	@Prop() inputWidth: '2-char-width' | '3-char-width' | '4-char-width' | '5-char-width' | '7-char-width' | '10-char-width' | '20-char-width' | 'default' = 'default';

	/**
	 * The name assigned to the input.The name value is used to reference form data after a form is submitted.
	 */
	@Prop() name: string;

	/**
	 * Used to define whether the input field is required or not. If required, the value passed should be 'required'.
	 */
	@Prop({ mutable: true }) required: boolean = false;

	/**
	 * The input type value.
	 */
	@Prop() type: 'text' | 'tel' | 'email' | 'password' = 'text';

	/**
	 * The input content value.
	 */
	@Prop({ mutable: true }) value?: string;

	/**
	 * Emitted when the input loses focus.
	 */
	@Event() blurEvent!: EventEmitter<void>;

	/**
	 * Emitted when the input gains focus.
	 */
	@Event() focusEvent!: EventEmitter<void>;

	/**
	 * Emitted when a keyboard input occurred.
	 */
	@Event() changeEvent!: EventEmitter<KeyboardEvent>;

	@State() focused: boolean = false;

	handleBlur = () => {
		this.focused = false;
	};

	handleFocus = () => {
		this.focused = true;
	};

	handleChange = (ev: Event) => {
		const input = ev.target as HTMLInputElement | null;

		if (input) {
			this.value = input.value ?? '';
		}

		this.changeEvent.emit(ev as KeyboardEvent);
	};

	private getValue(): string | number {
		return this.value ?? '';
	}

	private getClass(): string {
		return this.inputWidth === 'default' ? 'ontario-input' : `ontario-input ontario-input--${this.inputWidth}`;
	}

	public getId(): string {
		return this.elementId ?? '';
	}

	componentWillLoad() {
		this.elementId = this.elementId ?? uuid();
	}

	render() {
		return (
			<div>
				{getLabelElement(this.labelType, this.getId(), this.required, this.labelCaption)}
				<slot name="hint-text"></slot>
				<input
					aria-describedby={this.describedBy}
					class={this.getClass()}
					id={this.getId()}
					name={this.name}
					onBlur={this.handleBlur}
					onFocus={this.handleFocus}
					onInput={this.handleChange}
					required={this.required}
					type={this.type}
					value={this.getValue()}
				/>
				<slot name="hint-expander"></slot>
			</div>
		);
	}
}
