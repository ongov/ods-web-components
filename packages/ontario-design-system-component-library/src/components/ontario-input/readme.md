# ontario-input

The ontario-input is an interactive element often used in forms to capture data from the user.

## User Interaction Behaviour

User enters data into the input component.

## Usage guidance

Please refer to the [Ontario Design System](https://designsystem.ontario.ca) for current documentation guidance.

## Examples

Example of a bare-bones input component, the `name` property is passed in for the input, which is used to reference form data after a form is submitted. The `label-caption` property is being provided, which will be displayed as the label. By default, the `label-type` will be `default`, the `input-type` will be `text`, and the input will be set to `optional`.

```html
<ontario-input name="ontario-text-input" label-caption="Ontario input"/></ontario-input>
```

The following input component has the `input-width` property passed in, which will define the width of the input field. The `described-by` property has been passed in because the input has hint text associated with it.

```html
<ontario-input name="ontario-text-input" label-caption="Ontario input" input-width="10-char-width" describedBy="ontario-input-label-example" /><ontario-input></ontario-input>
```

## Accessibility

- An `id` attribute is necessary to allow the input to be associated with a label element
- A `name` attribute needs to be set to be submitted to the server when the form is submitted.

<!-- Auto Generated Below -->

## Properties

<<<<<<< HEAD
| Property | Attribute | Description | Type | Default |
| ------------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `caption` | `caption` | The text to display as the label | `InputCaption \| string` | `undefined` |
| `describedBy` | `described-by` | The aria-describedBy value if the input has hint text associated with it. | `string \| undefined` | `undefined` |
| `elementId` | `element-id` | The unique identifier of the input. If no ID is passed, one will be autogenerated. | `string \| undefined` | `undefined` |
| `hintText` | `hint-text` | Define hint text for Ontario input. | `string \| undefined` | `undefined` |
| `inputWidth` | `input-width` | The width of the input field. If no value is assigned, it will present as the default input width. | `"10-char-width" \| "2-char-width" \| "20-char-width" \| "3-char-width" \| "4-char-width" \| "5-char-width" \| "7-char-width" \| "default"` | `'default'` |
| `language` | `language` | The language of the component. This is used for translations, and is by default set through event listeners checking for a language property from the header. If none is passed, it will default to English. | `string \| undefined` | `'en'` |
| `name` | `name` | The name assigned to the input.The name value is used to reference form data after a form is submitted. | `string` | `undefined` |
| `required` | `required` | Used to define whether the input field is required or not. | `boolean` | `false` |
| `type` | `type` | The input type value. | `"email" \| "password" \| "tel" \| "text"` | `'text'` |
| `value` | `value` | The input content value. | `string \| undefined` | `undefined` |
=======
| Property | Attribute | Description | Type | Default |
| -------------- | --------------- | -------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `describedBy` | `described-by` | The aria-describedBy value if the input has hint text associated with it. | `string \| undefined` | `undefined` |
| `elementId` | `element-id` | The unique identifier of the input. If no ID is passed, one will be autogenerated. | `string \| undefined` | `undefined` |
| `hintExpander` | `hint-expander` | Used to include the Hint Expander component underneath the input box. This is passed in as an object with key-value pairs. | `HintExpander \| string \| undefined` | `undefined` |
| `inputWidth` | `input-width` | The width of the input field. If no value is assigned, it will present as the default input width. | `"10-char-width" \| "2-char-width" \| "20-char-width" \| "3-char-width" \| "4-char-width" \| "5-char-width" \| "7-char-width" \| "default"` | `'default'` |
| `labelCaption` | `label-caption` | The text to display as label. | `string` | `undefined` |
| `labelFor` | `label-for` | The form control with which the caption is associated. | `string \| undefined` | `undefined` |
| `labelType` | `label-type` | The type of label to render. | `"default" \| "heading" \| "large"` | `'default'` |
| `name` | `name` | The name assigned to the input.The name value is used to reference form data after a form is submitted. | `string` | `undefined` |
| `required` | `required` | Used to define whether the input field is required or not. | `boolean` | `false` |
| `type` | `type` | The input type value. | `"email" \| "password" \| "tel" \| "text"` | `'text'` |
| `value` | `value` | The input content value. | `string \| undefined` | `undefined` |

> > > > > > > 5d819109 (fixed merge conflicts)

## Events

| Event         | Description                             | Type                         |
| ------------- | --------------------------------------- | ---------------------------- |
| `blurEvent`   | Emitted when the input loses focus.     | `CustomEvent<void>`          |
| `changeEvent` | Emitted when a keyboard input occurred. | `CustomEvent<KeyboardEvent>` |
| `focusEvent`  | Emitted when the input gains focus.     | `CustomEvent<void>`          |

## Dependencies

### Depends on

- [ontario-hint-text](../ontario-hint-text)

### Graph

```mermaid
graph TD;
  ontario-input --> ontario-hint-text
  style ontario-input fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_