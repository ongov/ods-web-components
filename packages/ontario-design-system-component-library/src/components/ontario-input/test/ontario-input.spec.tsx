import { newSpecPage } from '@stencil/core/testing';
import { OntarioInput } from '../ontario-input';

describe('ontario-input', () => {
	describe('snapshot', () => {
		it('should render the expected html', async () => {
			const page = await newSpecPage({
				components: [OntarioInput],
				html: `<ontario-input element-id="ontario-input" caption='{"captionText": "Ontario Input"}' name="ontario-input"></ontario-input>`,
			});

			expect(page.root).toMatchSnapshot();
		});
	});

	describe('render', () => {
		it('should render a default input element', async () => {
			const page = await newSpecPage({
				components: [OntarioInput],
				html: `<ontario-input element-id="ontario-input" caption='{"captionText": "Ontario Input"}' name="ontario-input"></ontario-input>`,
			});
			expect(page.root).toEqualHtml(`
				<ontario-input element-id="ontario-input" caption='{"captionText": "Ontario Input"}' name="ontario-input">
					<mock:shadow-root>
						<div>
							<label class="ontario-label" htmlfor="ontario-input">
								Ontario Input
								<span class="ontario-label__flag">
									(optional)
								</span>
							</label>
							<slot name="hint-text"></slot>
							<input type='text' class="ontario-input" id="ontario-input" name="ontario-input" value=""></input>
							<slot name="hint-expander"></slot>
						</div>
					</mock:shadow-root>
				</ontario-input>
			`);
		});

		it('should reflect attributes/props being set', async () => {
			const page = await newSpecPage({
				components: [OntarioInput],
				html: `<ontario-input
									described-by="input-hint-id"
									name="input-name"
									element-id="input-id"
									value="input value"
									type='tel'
									input-width='7-char-width'
									caption='{"captionText": "Ontario Input", "isRequired": true}'
								></ontario-input>`,
			});

			expect(page.rootInstance.describedBy).toBe('input-hint-id');
			expect(page.rootInstance.name).toBe('input-name');
			expect(page.rootInstance.captionState.isRequired).toBe(true);
			expect(page.rootInstance.value).toBe('input value');
			expect(page.rootInstance.elementId).toBe('input-id');
			expect(page.rootInstance.type).toBe('tel');
			expect(page.rootInstance.inputWidth).toBe('7-char-width');
			expect(page.rootInstance.focused).toBe(false);
			expect(page.rootInstance.captionState.captionText).toBe('Ontario Input');
		});
	});

	describe('events/methods', () => {
		it('should handle focus with right focused state', async () => {
			const input = new OntarioInput();
			input.handleFocus();
			expect(input.focused).toBeTruthy();
		});

		it('should handle blur with right focused state', async () => {
			const input = new OntarioInput();
			input.focused = true;
			input.handleBlur();
			expect(input.focused).toBeFalsy();
		});

		it('should emit a keyboard event on change', async () => {
			const page = await newSpecPage({
				components: [OntarioInput],
				html: `<ontario-input
									name="input-name"
									element-id="input-id"
								></ontario-input>`,
			});

			const emitSpy = jest.fn();
			const leftArrowKeyCode = 37;
			page.doc.addEventListener('changeEvent', emitSpy);
			page.rootInstance.handleChange(
				new KeyboardEvent('keydown', {
					keyCode: leftArrowKeyCode,
				}),
			);
			await page.waitForChanges();
			expect(emitSpy).toHaveBeenCalled();
		});

		it('should update the input value on a change event', async () => {
			const page = await newSpecPage({
				components: [OntarioInput],
				html: `<ontario-input
									name="input-name"
									element-id="input-id"
								></ontario-input>`,
			});

			const emitSpy = jest.fn();
			const testValue = 'This is a test';
			const leftArrowKeyCode = 37;
			page.doc.addEventListener('changeEvent', emitSpy);
			page.rootInstance.value = testValue;
			page.rootInstance.handleChange(
				new KeyboardEvent('keydown', {
					keyCode: leftArrowKeyCode,
				}),
			);
			await page.waitForChanges();
			expect(page.rootInstance.value).toBe(testValue);
		});

		it('should return the input id when using the getId method', async () => {
			const page = await newSpecPage({
				components: [OntarioInput],
				html: `<ontario-input
									name="input-name"
									element-id="input-id"
								></ontario-input>`,
			});

			expect(page.rootInstance.getId()).toEqual('input-id');
		});
	});
});
