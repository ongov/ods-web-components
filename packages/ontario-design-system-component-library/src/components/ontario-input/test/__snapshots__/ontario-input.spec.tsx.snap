// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-input snapshot should render the expected html 1`] = `
<ontario-input caption="{&quot;captionText&quot;: &quot;Ontario Input&quot;}" element-id="ontario-input" name="ontario-input">
  <mock:shadow-root>
    <div>
      <label class="ontario-label" htmlfor="ontario-input">
        Ontario Input
        <span class="ontario-label__flag">
          (optional)
        </span>
      </label>
      <slot name="hint-text"></slot>
      <input class="ontario-input ontario-input-hint-expander--false" id="ontario-input" name="ontario-input" type="text" value="">
    </div>
  </mock:shadow-root>
</ontario-input>
`;
